heron/api/src/java/com/twitter/heron/api/state/State.java

```diff
@@ -0,0 +1,26 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.api.state;
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * State represents the state interface as seen by stateful bolts and spouts.
+ * In Heron, state gives a notional Key/Value interface along with the
+ * ability to iterate over the key/values
+ */
+public interface State extends Map<Serializable, Serializable> {
```

Comment:
yes, but also if we had K, V users could write this:
```
  class MyStatefulBolt implements ITypedStatefulComponent<String, Integer> {
    @Override
    public void initState(TypedState<String, Integer> state) {

    }
  }
```
and work with the types that it expects, just like they would while working with a typed map like `Map<String, Integer>` (instead of `Map<Serializable, Serializable>`. This provides compile time checks in user code for their types and removed the runtime need to cast objects, which leads to more bug-free and easy to maintain code.

I would need to prototype though to see that we could in fact create an instance of `TypedState<String, Integer>` at runtime though, give persisted `Serializables` and an instance of `MyStatefulBolt<String, Integer>` to derive types from.